The job **`mingw-check-tidy`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/10750481082/job/29816684612) [(plain)](https://github.com/rust-lang/rust/commit/9ac24800371f158320e128d03b6ee66ba55b7ce4/checks/29816684612/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain

COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
COPY host-x86_64/mingw-check/validate-error-codes.sh /scripts/

# NOTE: intentionally uses python2 for x.py so we can test it still works.
# validate-toolstate only runs in our CI, so it's ok for it to only support python3.
ENV SCRIPT TIDY_PRINT_DIFF=1 python2.7 ../x.py test \
           --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint,cpp:fmt
# This file is autogenerated by pip-compile with Python 3.10
# by the following command:
#
#    pip-compile --allow-unsafe --generate-hashes reuse-requirements.in
---
#13 2.821 Building wheels for collected packages: reuse
#13 2.822   Building wheel for reuse (pyproject.toml): started
#13 3.066   Building wheel for reuse (pyproject.toml): finished with status 'done'
#13 3.067   Created wheel for reuse: filename=reuse-4.0.3-cp310-cp310-manylinux_2_35_x86_64.whl size=132715 sha256=dfa09868353292d98f811d3efdb0d54d07389e808efc71d68e3b93c514bf8bec
#13 3.067   Stored in directory: /tmp/pip-ephem-wheel-cache-s4ltlj3z/wheels/3d/8d/0a/e0fc6aba4494b28a967ab5eaf951c121d9c677958714e34532
#13 3.070 Installing collected packages: boolean-py, binaryornot, tomlkit, reuse, python-debian, markupsafe, license-expression, jinja2, chardet, attrs
#13 3.470 Successfully installed attrs-23.2.0 binaryornot-0.4.4 boolean-py-4.0 chardet-5.2.0 jinja2-3.1.4 license-expression-30.3.0 markupsafe-2.1.5 python-debian-0.1.49 reuse-4.0.3 tomlkit-0.13.0
#13 3.470 WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv
#13 3.996 Collecting virtualenv
#13 3.996 Collecting virtualenv
#13 4.046   Downloading virtualenv-20.26.3-py3-none-any.whl (5.7 MB)
#13 4.249      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 5.7/5.7 MB 28.2 MB/s eta 0:00:00
#13 4.313 Collecting filelock<4,>=3.12.2
#13 4.320   Downloading filelock-3.15.4-py3-none-any.whl (16 kB)
#13 4.340 Collecting distlib<1,>=0.3.7
#13 4.346   Downloading distlib-0.3.8-py2.py3-none-any.whl (468 kB)
#13 4.357      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 468.9/468.9 KB 56.1 MB/s eta 0:00:00
#13 4.389 Collecting platformdirs<5,>=3.9.1
#13 4.396   Downloading platformdirs-4.2.2-py3-none-any.whl (18 kB)
#13 4.476 Installing collected packages: distlib, platformdirs, filelock, virtualenv
#13 4.663 Successfully installed distlib-0.3.8 filelock-3.15.4 platformdirs-4.2.2 virtualenv-20.26.3
#13 DONE 4.7s

#14 [7/8] COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
#14 DONE 0.0s
---
DirectMap4k:      225216 kB
DirectMap2M:     7114752 kB
DirectMap1G:    11534336 kB
##[endgroup]
Executing TIDY_PRINT_DIFF=1 python2.7 ../x.py test            --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint,cpp:fmt
+ TIDY_PRINT_DIFF=1 python2.7 ../x.py test --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint,cpp:fmt
    Finished `dev` profile [unoptimized] target(s) in 0.04s
##[endgroup]
downloading https://ci-artifacts.rust-lang.org/rustc-builds-alt/9afe7136958edaa403f0b0eb00f0353c125b7352/rust-dev-nightly-x86_64-unknown-linux-gnu.tar.xz
extracting /checkout/obj/build/cache/llvm-9afe7136958edaa403f0b0eb00f0353c125b7352-true/rust-dev-nightly-x86_64-unknown-linux-gnu.tar.xz to /checkout/obj/build/x86_64-unknown-linux-gnu/ci-llvm
---
##[endgroup]
fmt check
Diff in /checkout/compiler/rustc_target/src/spec/base/apple/mod.rs:190:
         //
         // CC forwards the `-arch` to the linker, so we use the same value
         // here intentionally.
-        add_link_args(&mut args, LinkerFlavor::Darwin(Cc::Yes, Lld::No), &["-arch", arch.ld_arch()]);
+            &mut args,
+            &mut args,
+            LinkerFlavor::Darwin(Cc::Yes, Lld::No),
+            &["-arch", arch.ld_arch()],
+        );
         // The presence of `-mmacosx-version-min` makes CC default to macOS,
         // and it sets the deployment target.
         let (major, minor, patch) = deployment_target(os, arch, abi);
fmt error: Running `"/checkout/obj/build/x86_64-unknown-linux-gnu/rustfmt/bin/rustfmt" "--config-path" "/checkout" "--edition" "2021" "--unstable-features" "--skip-children" "--check" "/checkout/compiler/rustc_target/src/spec/base/hermit.rs" "/checkout/compiler/rustc_target/src/spec/base/vxworks.rs" "/checkout/compiler/rustc_target/src/spec/base/solaris.rs" "/checkout/compiler/rustc_target/src/spec/base/xtensa.rs" "/checkout/compiler/rustc_target/src/spec/base/windows_gnullvm.rs" "/checkout/compiler/rustc_target/src/spec/base/wasm.rs" "/checkout/compiler/rustc_target/src/spec/base/unikraft_linux_musl.rs" "/checkout/compiler/rustc_target/src/spec/base/aix.rs" "/checkout/compiler/rustc_target/src/spec/base/teeos.rs" "/checkout/compiler/rustc_target/src/spec/base/uefi_msvc.rs" "/checkout/compiler/rustc_target/src/spec/base/freebsd.rs" "/checkout/compiler/rustc_target/src/spec/base/nto_qnx.rs" "/checkout/compiler/rustc_target/src/spec/base/avr_gnu.rs" "/checkout/compiler/rustc_target/src/spec/base/windows_uwp_msvc.rs" "/checkout/compiler/rustc_target/src/spec/base/bpf.rs" "/checkout/compiler/rustc_target/src/spec/base/linux_uclibc.rs" "/checkout/compiler/rustc_target/src/spec/base/linux_musl.rs" "/checkout/compiler/rustc_target/src/spec/base/linux.rs" "/checkout/compiler/rustc_target/src/spec/base/msvc.rs" "/checkout/compiler/rustc_target/src/spec/base/illumos.rs" "/checkout/compiler/rustc_target/src/spec/base/android.rs" "/checkout/compiler/rustc_target/src/spec/base/hurd_gnu.rs" "/checkout/compiler/rustc_target/src/spec/base/windows_uwp_gnu.rs" "/checkout/compiler/rustc_target/src/spec/base/haiku.rs" "/checkout/compiler/rustc_target/src/spec/base/dragonfly.rs" "/checkout/compiler/rustc_target/src/spec/base/windows_gnu.rs" "/checkout/compiler/rustc_target/src/spec/base/mod.rs" "/checkout/compiler/rustc_target/src/spec/base/windows_msvc.rs" "/checkout/compiler/rustc_target/src/spec/base/netbsd.rs" "/checkout/compiler/rustc_target/src/spec/base/linux_ohos.rs" "/checkout/compiler/rustc_target/src/spec/base/hurd.rs" "/checkout/compiler/rustc_target/src/spec/base/apple/tests.rs" "/checkout/compiler/rustc_target/src/spec/base/apple/mod.rs" "/checkout/compiler/rustc_target/src/spec/crt_objects.rs" "/checkout/compiler/rustc_target/src/spec/tests/tests_impl.rs" "/checkout/compiler/rustc_codegen_gcc/build_system/src/utils.rs" "/checkout/compiler/rustc_codegen_gcc/build_system/src/clean.rs" "/checkout/compiler/rustc_codegen_gcc/build_system/src/clone_gcc.rs" "/checkout/compiler/rustc_codegen_gcc/build_system/src/info.rs" "/checkout/compiler/rustc_codegen_gcc/build_system/src/build.rs" "/checkout/compiler/rustc_codegen_gcc/build_system/src/rustc_info.rs" "/checkout/compiler/rustc_codegen_gcc/build_system/src/config.rs" "/checkout/compiler/rustc_codegen_gcc/build_system/src/fmt.rs" "/checkout/compiler/rustc_codegen_gcc/build_system/src/prepare.rs" "/checkout/compiler/rustc_codegen_gcc/build_system/src/test.rs" "/checkout/compiler/rustc_codegen_gcc/build_system/src/rust_tools.rs" "/checkout/compiler/rustc_codegen_gcc/build_system/src/main.rs" "/checkout/compiler/rustc_codegen_gcc/build_system/build_sysroot/lib.rs" "/checkout/compiler/rustc_codegen_gcc/src/back/write.rs" "/checkout/compiler/rustc_codegen_gcc/src/back/mod.rs" "/checkout/compiler/rustc_codegen_gcc/src/back/lto.rs" "/checkout/compiler/rustc_codegen_gcc/src/intrinsic/llvm.rs" "/checkout/compiler/rustc_codegen_gcc/src/intrinsic/simd.rs" "/checkout/compiler/rustc_codegen_gcc/src/intrinsic/mod.rs" "/checkout/compiler/rustc_codegen_gcc/src/callee.rs" "/checkout/compiler/rustc_codegen_gcc/src/abi.rs" "/checkout/compiler/rustc_codegen_gcc/src/builder.rs" "/checkout/compiler/rustc_codegen_gcc/src/debuginfo.rs" "/checkout/compiler/rustc_codegen_gcc/src/archive.rs" "/checkout/compiler/rustc_codegen_gcc/src/gcc_util.rs" "/checkout/compiler/rustc_codegen_gcc/src/consts.rs" "/checkout/compiler/rustc_codegen_gcc/src/allocator.rs" "/checkout/compiler/rustc_codegen_gcc/src/attributes.rs" "/checkout/compiler/rustc_target/src/spec/base/solid.rs"` failed.
If you're running `tidy`, try again with `--bless`. Or, if you just want to format code, run `./x.py fmt` instead.
  local time: Sat Sep  7 09:57:33 UTC 2024
  network time: Sat, 07 Sep 2024 09:57:33 GMT
##[error]Process completed with exit code 1.
Post job cleanup.
```

</details>

--- a/compiler/rustc_target/src/spec/base/apple/mod.rs
+++ b/compiler/rustc_target/src/spec/base/apple/mod.rs
@@ -168,7 +168,26 @@ fn pre_link_args(os: &'static str, arch: Arch, abi: TargetAbi) -> LinkArgs {
         ["-platform_version".into(), platform_name, min_version, sdk_version].into_iter(),
     );
 
-    if abi != TargetAbi::MacCatalyst {
+    // We need to communicate four things to the C compiler to be able to link:
+    // - The architecture.
+    // - The operating system (and that it's an Apple platform).
+    // - The deployment target.
+    // - The environment / ABI.
+    //
+    // We'd like to use `-target` everywhere, since that can uniquely
+    // communicate all of these, but that doesn't work on GCC, and since we
+    // don't know whether the `cc` compiler is Clang, GCC, or something else,
+    // we fall back to other options that also work on GCC when compiling for
+    // macOS.
+    //
+    // Targets other than macOS are ill-supported by GCC (it doesn't even
+    // support e.g. `-miphoneos-version-min`), so in those cases we can fairly
+    // safely use `-target`. See also the following, where it is made explicit
+    // that the recommendation by LLVM developers is to use `-target`:
+    // <https://github.com/llvm/llvm-project/issues/88271>
+    if os == "macos" {
+        // `-arch` communicates the architecture.
+        //
         // CC forwards the `-arch` to the linker, so we use the same value
         // here intentionally.
         add_link_args(
@@ -176,6 +195,15 @@ fn pre_link_args(os: &'static str, arch: Arch, abi: TargetAbi) -> LinkArgs {
             LinkerFlavor::Darwin(Cc::Yes, Lld::No),
             &["-arch", arch.ld_arch()],
         );
+        // The presence of `-mmacosx-version-min` makes CC default to macOS,
+        // and it sets the deployment target.
+        let (major, minor, patch) = deployment_target(os, arch, abi);
+        let opt = format!("-mmacosx-version-min={major}.{minor}.{patch}").into();
+        add_link_args_iter(&mut args, LinkerFlavor::Darwin(Cc::Yes, Lld::No), [opt].into_iter());
+        // macOS has no environment, so with these two, we've told CC all the
+        // desired parameters.
+        //
+        // We avoid `-m32`/`-m64`, as this is already encoded by `-arch`.
     } else {
         add_link_args_iter(
             &mut args,

--- a/compiler/rustc_target/src/spec/targets/i686_apple_darwin.rs
+++ b/compiler/rustc_target/src/spec/targets/i686_apple_darwin.rs
@@ -1,10 +1,8 @@
 use crate::spec::base::apple::{base, Arch, TargetAbi};
-use crate::spec::{Cc, FramePointer, LinkerFlavor, Lld, Target, TargetOptions};
+use crate::spec::{FramePointer, Target, TargetOptions};
 
 pub(crate) fn target() -> Target {
-    let (mut opts, llvm_target, arch) = base("macos", Arch::I686, TargetAbi::Normal);
-    opts.add_pre_link_args(LinkerFlavor::Darwin(Cc::Yes, Lld::No), &["-m32"]);
-
+    let (opts, llvm_target, arch) = base("macos", Arch::I686, TargetAbi::Normal);
     Target {
         llvm_target,
         metadata: crate::spec::TargetMetadata {

--- a/compiler/rustc_target/src/spec/targets/x86_64_apple_darwin.rs
+++ b/compiler/rustc_target/src/spec/targets/x86_64_apple_darwin.rs
@@ -1,9 +1,8 @@
 use crate::spec::base::apple::{base, Arch, TargetAbi};
-use crate::spec::{Cc, FramePointer, LinkerFlavor, Lld, SanitizerSet, Target, TargetOptions};
+use crate::spec::{FramePointer, SanitizerSet, Target, TargetOptions};
 
 pub(crate) fn target() -> Target {
-    let (mut opts, llvm_target, arch) = base("macos", Arch::X86_64, TargetAbi::Normal);
-    opts.add_pre_link_args(LinkerFlavor::Darwin(Cc::Yes, Lld::No), &["-m64"]);
+    let (opts, llvm_target, arch) = base("macos", Arch::X86_64, TargetAbi::Normal);
     Target {
         llvm_target,
         metadata: crate::spec::TargetMetadata {

--- a/compiler/rustc_target/src/spec/targets/x86_64h_apple_darwin.rs
+++ b/compiler/rustc_target/src/spec/targets/x86_64h_apple_darwin.rs
@@ -1,11 +1,10 @@
 use crate::spec::base::apple::{base, Arch, TargetAbi};
-use crate::spec::{Cc, FramePointer, LinkerFlavor, Lld, SanitizerSet, Target, TargetOptions};
+use crate::spec::{FramePointer, SanitizerSet, Target, TargetOptions};
 
 pub(crate) fn target() -> Target {
     let (mut opts, llvm_target, arch) = base("macos", Arch::X86_64h, TargetAbi::Normal);
     opts.max_atomic_width = Some(128);
     opts.frame_pointer = FramePointer::Always;
-    opts.add_pre_link_args(LinkerFlavor::Darwin(Cc::Yes, Lld::No), &["-m64"]);
     opts.supported_sanitizers =
         SanitizerSet::ADDRESS | SanitizerSet::CFI | SanitizerSet::LEAK | SanitizerSet::THREAD;
 

--- a/tests/run-make/apple-deployment-target/rmake.rs
+++ b/tests/run-make/apple-deployment-target/rmake.rs
@@ -81,9 +81,8 @@ fn main() {
         rustc().env(env_var, example_version).run();
         minos("libfoo.dylib", example_version);
 
-        // FIXME(madsmtm): Deployment target is not currently passed properly to linker
-        // rustc().env_remove(env_var).run();
-        // minos("libfoo.dylib", default_version);
+        rustc().env_remove(env_var).run();
+        minos("libfoo.dylib", default_version);
 
         // Test with ld64 instead
 
@@ -110,9 +109,8 @@ fn main() {
             rustc().env(env_var, example_version).run();
             minos("foo", example_version);
 
-            // FIXME(madsmtm): Deployment target is not currently passed properly to linker
-            // rustc().env_remove(env_var).run();
-            // minos("foo", default_version);
+            rustc().env_remove(env_var).run();
+            minos("foo", default_version);
         }
 
         // Test with ld64 instead

